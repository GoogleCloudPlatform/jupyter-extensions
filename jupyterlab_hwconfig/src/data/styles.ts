/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as csstips from 'csstips';
import { stylesheet } from 'typestyle';
import { BASE_FONT } from 'gcp_jupyterlab_shared';

/* Class names applied to the component. */
export const STYLES = stylesheet({
  container: {
    color: 'var(--jp-ui-font-color1)',
    fontFamily: 'var(--jp-ui-font-family)',
    fontSize: 'var(--jp-ui-font-size1, 13px)',
    lineHeight: '24px',
    alignItems: 'center',
    ...csstips.horizontal,
  },
  containerPadding: {
    margin: '0px',
    padding: '35px 30px',
    backgroundColor: 'var(--jp-layout-color1)',
  },
  containerSize: {
    width: 468,
  },
  heading: {
    fontSize: '22px',
    paddingBottom: '10px',
    fontWeight: 400,
    fontFamily: 'var(--jp-ui-font-family)',
    color: 'var(--jp-ui-font-color1)',
    display: 'block',
  },
  subheading: {
    fontFamily: 'var(--jp-ui-font-family)',
    color: 'var(--jp-ui-font-color1)',
    fontWeight: 700,
    fontSize: '15px',
    paddingTop: '20px',
    paddingBottom: '5px',
    display: 'block',
  },
  paragraph: {
    fontSize: '14px',
    paddingTop: '2px',
    display: 'block',
    fontWeight: 400,
    fontFamily: 'var(--jp-ui-font-family)',
    color: 'var(--jp-ui-font-color1)',
  },
  attribute: {
    marginRight: '4px',
  },
  interactiveHover: {
    $nest: {
      '&:hover': {
        backgroundColor: '#8a8a8a',
      },
    },
  },
  icon: {
    display: 'inline-block',
    height: '18px',
    marginRight: '4px',
    width: '18px',
  },
  listRow: {
    display: 'table-row',
    boxShadow: 'inset 0 -1px 0 0 var(--jp-border-color1)',
  },
  infoMessage: {
    marginTop: '20px',
  },
});

export const TEXT_STYLE = {
  fontFamily: BASE_FONT.fontFamily as string,
  fontSize: BASE_FONT.fontSize as number,
  color: 'var(--jp-ui-font-color1)',
};
