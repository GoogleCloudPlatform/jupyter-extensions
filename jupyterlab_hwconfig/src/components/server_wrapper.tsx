/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ServerConnection } from '@jupyterlab/services';
import { Details } from '../data/data';

export class ServerWrapper {
  private readonly serverSettings: ServerConnection.ISettings;
  private readonly url: string;
  constructor(
    serverSettings: ServerConnection.ISettings = ServerConnection.defaultSettings
  ) {
    this.serverSettings = serverSettings;
    this.url = this.serverSettings.baseUrl + `gcp/v1/details`;
  }

  async getUtilizationData(): Promise<Details> {
    const response = await ServerConnection.makeRequest(
      this.url,
      {},
      this.serverSettings
    );
    if (!response.ok) {
      throw new Error();
    }
    const data = await response.json();
    return { ...data };
  }
}
