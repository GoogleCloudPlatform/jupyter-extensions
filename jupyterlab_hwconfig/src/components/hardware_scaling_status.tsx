/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as csstips from 'csstips';
import * as React from 'react';
import { stylesheet, classes } from 'typestyle';
import LinearProgress from '@material-ui/core/LinearProgress';
import { createStyles, withStyles, Theme } from '@material-ui/core/styles';
import { STYLES } from '../data/styles';
import { HardwareConfiguration } from '../data/data';
import { MachineTypeConfiguration } from '../data/machine_types';
import { ActionBar } from './action_bar';
import { NotebooksService, Instance } from '../service/notebooks_service';
import { ServerWrapper } from './server_wrapper';
import { ErrorPage } from './error_page';
import { displayInstance } from './instance_details_message';

const BorderLinearProgress = withStyles((theme: Theme) =>
  createStyles({
    root: {
      height: 10,
      width: 450,
      borderRadius: 15,
      margin: '10px 0px',
    },
    colorPrimary: {
      backgroundColor:
        theme.palette.grey[theme.palette.type === 'light' ? 200 : 700],
    },
    bar: {
      backgroundColor: '#1a90ff',
      borderRadius: 5,
    },
  })
)(LinearProgress);

const STATUS_STYLES = stylesheet({
  container: {
    ...csstips.vertical,
    alignItems: 'center',
    backgroundColor: 'var(--jp-layout-color1)',
  },
  containerPadding: {
    padding: '50px 75px 15px 75px',
  },
  containerBottomPadding: {
    paddingBottom: '20px',
  },
  image: {
    marginBottom: '40px',
    height: 240,
    width: 300,
  },
  bottomText: {
    marginTop: '50px',
    fontSize: '12px',
    color: 'var(--jp-ui-font-color2)',
  },
});

export enum Status {
  'Authorizing' = 0,
  'Stopping notebook instance' = 1,
  'Updating machine configuration' = 2,
  'Updating GPU configuration' = 3,
  'Restarting notebook instance' = 4,
  'Refreshing session' = 5,
  'Complete' = 6,
  'Error' = 7,
}

const statusInfo = [
  'Please complete the OAuth 2.0 authorization steps in the popup',
  'Shutting down your notebook instance for reshaping.',
  'Reshaping machine configuration to match your selection.',
  'Reshaping GPU configuration to match your selection.',
  'Restarting your newly configured notebook instance.',
  'Refreshing your JupyterLab session to reflect your new configuration.',
  'Operation complete. Enjoy your newly configured instance!',
  'An error has occured, please try again later.',
];

export enum ErrorType {
  STOP = 'Stop',
  RESHAPING = 'Reshape',
  START = 'Start',
  REFRESH = 'Refresh',
}

export interface ConfigurationError {
  errorType: ErrorType;
  errorValue: any;
}

interface Props {
  hardwareConfiguration: HardwareConfiguration;
  notebookService: NotebooksService;
  onDialogClose: () => void;
  onCompletion: () => void;
  detailsServer: ServerWrapper;
  authTokenRetrieval: () => Promise<string>;
  machineTypes: MachineTypeConfiguration[];
}

interface State {
  status: Status;
  instanceDetails: Instance;
  error: ConfigurationError;
}

export class HardwareScalingStatus extends React.Component<Props, State> {
  private readonly NUM_RETRIES = 20;

  constructor(props: Props) {
    super(props);
    this.state = {
      status: Status.Authorizing,
      instanceDetails: null,
      error: null,
    };
  }

  private preventPageClose(event) {
    event.preventDefault();
    event.returnValue = '';
  }

  private updateStatus(status: Status, instanceDetails: Instance) {
    this.setState({
      status,
      instanceDetails,
    });
  }

  private showError(
    errorType: ErrorType,
    errorValue: any,
    newInstanceDetails?: Instance
  ) {
    this.setState({
      status: Status.Error,
      instanceDetails: newInstanceDetails
        ? newInstanceDetails
        : this.state.instanceDetails,
      error: {
        errorType,
        errorValue,
      },
    });
  }

  private updateError(
    errorType: ErrorType,
    errorValue: any,
    newInstanceDetails?: Instance
  ) {
    this.setState({
      instanceDetails: newInstanceDetails
        ? newInstanceDetails
        : this.state.instanceDetails,
      error: {
        errorType,
        errorValue,
      },
    });
  }

  private async waitForServer() {
    for (let tries = 0; tries < this.NUM_RETRIES; tries++) {
      try {
        await this.props.detailsServer.getUtilizationData();
        break;
      } catch (err) {
        if (tries === this.NUM_RETRIES - 1) {
          this.showError(ErrorType.REFRESH, err);
        }
        await new Promise(resolve => setTimeout(resolve, 5000));
      }
    }
  }

  async componentDidMount() {
    const { notebookService, authTokenRetrieval } = this.props;
    try {
      const token = await authTokenRetrieval();
      this.setState({
        status: Status['Stopping notebook instance'],
      });
      notebookService.setAuthToken(token);
    } catch (err) {
      this.setState({ status: Status.Error });
      console.log(err);
    }
    window.addEventListener('beforeunload', this.preventPageClose);
  }

  async componentDidUpdate(prevProps, prevState) {
    const { status, error, instanceDetails } = this.state;
    const { hardwareConfiguration, notebookService, onCompletion } = this.props;
    const { machineType, attachGpu, gpuType, gpuCount } = hardwareConfiguration;

    if (prevState.status !== status) {
      switch (status) {
        case Status['Stopping notebook instance']: {
          try {
            const stopResult = await notebookService.stop();
            this.updateStatus(
              Status['Updating machine configuration'],
              stopResult
            );
          } catch (err) {
            this.showError(ErrorType.STOP, err);
          }
          break;
        }

        case Status['Updating machine configuration']: {
          let setMachineTypeResult = instanceDetails;
          try {
            setMachineTypeResult = await notebookService.setMachineType(
              machineType.name
            );
          } catch (err) {
            this.updateError(ErrorType.RESHAPING, err);
          }
          const nextStatus = attachGpu
            ? 'Updating GPU configuration'
            : 'Restarting notebook instance';
          this.updateStatus(Status[nextStatus], setMachineTypeResult);
          break;
        }

        case Status['Updating GPU configuration']: {
          let setGpuTypeResult = instanceDetails;
          try {
            setGpuTypeResult = await notebookService.setAccelerator(
              gpuType,
              gpuCount
            );
          } catch (err) {
            this.updateError(ErrorType.RESHAPING, err);
          }
          this.updateStatus(
            Status['Restarting notebook instance'],
            setGpuTypeResult
          );
          break;
        }

        case Status['Restarting notebook instance']: {
          try {
            const startResult = await notebookService.start();
            this.updateStatus(Status['Refreshing session'], startResult);
          } catch (err) {
            this.showError(ErrorType.START, err);
          }
          break;
        }

        case Status['Refreshing session']: {
          await this.waitForServer();
          onCompletion();
          this.setState({
            status:
              error && error.errorType === ErrorType.RESHAPING
                ? Status.Error
                : Status.Complete,
          });
          break;
        }

        default:
          break;
      }
    }
  }

  componentWillUnmount() {
    window.removeEventListener('beforeunload', this.preventPageClose);
  }

  render() {
    const { status, error, instanceDetails } = this.state;
    const progressValue = ((status + 1) / 6) * 100;
    const { onDialogClose, machineTypes } = this.props;
    return status === Status['Error'] ? (
      <ErrorPage
        onDialogClose={onDialogClose}
        error={error}
        instanceDetails={instanceDetails}
        machineTypes={machineTypes}
      />
    ) : (
      <div
        className={classes(
          STATUS_STYLES.containerPadding,
          STATUS_STYLES.container
        )}
      >
        <span className={classes(STATUS_STYLES.image, 'jp-CaipIcon')} />
        <p className={STYLES.heading}>
          {status < 6 ? `${Status[status]} (${status + 1}/6)` : Status[status]}
        </p>
        {status === Status['Complete'] ? (
          <div
            className={
              (STATUS_STYLES.container, STATUS_STYLES.containerBottomPadding)
            }
          >
            <p className={STYLES.paragraph}>{statusInfo[status]}</p>
            {status === Status['Complete'] &&
              displayInstance(
                instanceDetails,
                machineTypes,
                'Your new configuration:'
              )}
            <ActionBar onPrimaryClick={onDialogClose} primaryLabel="Close" />
          </div>
        ) : (
          <div className={STATUS_STYLES.container}>
            <BorderLinearProgress variant="determinate" value={progressValue} />
            <p className={STYLES.paragraph}>{statusInfo[status]}</p>
            <p className={STATUS_STYLES.bottomText}>
              Don't close the browser tab before the update process is finished
            </p>
          </div>
        )}
      </div>
    );
  }
}
