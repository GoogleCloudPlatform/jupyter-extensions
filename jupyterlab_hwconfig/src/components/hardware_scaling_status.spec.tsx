/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* eslint-disable @typescript-eslint/no-empty-function */

import * as React from 'react';
import { shallow } from 'enzyme';
import { HardwareScalingStatus, Status } from './hardware_scaling_status';
import { DETAILS_RESPONSE } from '../test_helpers';
import { NotebooksService, Instance } from '../service/notebooks_service';
import { ServerWrapper } from './server_wrapper';
import { detailsToHardwareConfiguration } from '../data/data';
import { MachineTypeConfiguration } from '../data/machine_types';

function immediatePromise() {
  return new Promise(r => setTimeout(r));
}

describe('HardwareScalingStatus', () => {
  const mockGetUtilizationData = jest.fn();
  const mockStopInstance = jest.fn();
  const mockSetMachineType = jest.fn();
  const mockSetAccelerator = jest.fn();
  const mockStartInstance = jest.fn();
  const mockAuthTokenRetrieval = jest.fn();
  const mockSetAuthToken = jest.fn();
  const mockOnComplete = jest.fn();
  const mockOnDialogClose = jest.fn();
  const mockServerWrapper = ({
    getUtilizationData: mockGetUtilizationData,
  } as unknown) as ServerWrapper;
  const mockNotebookService = ({
    stop: mockStopInstance,
    setMachineType: mockSetMachineType,
    setAccelerator: mockSetAccelerator,
    start: mockStartInstance,
    setAuthToken: mockSetAuthToken,
  } as unknown) as NotebooksService;
  const mockInstance: Instance = {
    name: 'Test',
    machineType:
      'https://www.googleapis.com/compute/v1/projects/jupyterlab-interns-sandbox/zones/us-west1-b/machineTypes/e2-highmem-8',
    acceleratorConfig: {
      coreCount: '1',
      type: 'NVIDIA_TESLA_K80',
    },
  };
  const mockToken = 'mockToken';
  const mockMachineTypes: MachineTypeConfiguration[] = [
    {
      base: { value: 'e2-', text: 'Efficient Instance' },
      configurations: [
        {
          value: 'e2-highmem-8',
          text: 'Efficient Instance, 8 vCPUs, 64 GB RAM',
        },
      ],
    },
  ];
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('Runs through reshaping flow', async () => {
    mockAuthTokenRetrieval.mockResolvedValue(mockToken);
    mockStopInstance.mockResolvedValue(mockInstance);
    mockSetMachineType.mockResolvedValue(mockInstance);
    mockSetAccelerator.mockResolvedValue(mockInstance);
    mockStartInstance.mockResolvedValue(mockInstance);
    const detailsResponse = await JSON.parse(DETAILS_RESPONSE);
    const resolveValue = Promise.resolve({ ...detailsResponse });
    mockGetUtilizationData.mockReturnValue(resolveValue);
    const hardwareConfig = detailsToHardwareConfiguration(detailsResponse);
    const hardwareScalingStatus = shallow(
      <HardwareScalingStatus
        detailsServer={mockServerWrapper}
        notebookService={mockNotebookService}
        onCompletion={mockOnComplete}
        onDialogClose={mockOnDialogClose}
        hardwareConfiguration={hardwareConfig}
        authTokenRetrieval={mockAuthTokenRetrieval}
        machineTypes={mockMachineTypes}
      />
    );
    expect(hardwareScalingStatus.state('status')).toEqual(Status.Authorizing);
    expect(hardwareScalingStatus).toMatchSnapshot('Authorizing');
    expect(mockAuthTokenRetrieval).toBeCalled();
    await mockAuthTokenRetrieval();
    expect(mockNotebookService.setAuthToken).toHaveBeenCalledWith(mockToken);
    expect(hardwareScalingStatus.state('status')).toEqual(
      Status['Stopping notebook instance']
    );
    expect(hardwareScalingStatus).toMatchSnapshot('Stopping notebook instance');
    expect(mockNotebookService.stop).toBeCalled();
    await mockStopInstance();
    expect(hardwareScalingStatus.state('status')).toEqual(
      Status['Updating machine configuration']
    );
    expect(hardwareScalingStatus).toMatchSnapshot(
      'Updating machine configuration'
    );
    expect(mockNotebookService.setMachineType).toBeCalled();
    await mockSetMachineType();
    expect(hardwareScalingStatus.state('status')).toEqual(
      Status['Updating GPU configuration']
    );
    expect(hardwareScalingStatus).toMatchSnapshot('Updating GPU configuration');
    expect(mockNotebookService.setAccelerator).toBeCalled();
    await mockSetAccelerator();
    expect(hardwareScalingStatus.state('status')).toEqual(
      Status['Restarting notebook instance']
    );
    expect(hardwareScalingStatus).toMatchSnapshot(
      'Restarting notebook instance'
    );
    expect(mockNotebookService.start).toBeCalled();
    await mockStartInstance();
    expect(hardwareScalingStatus.state('status')).toEqual(
      Status['Refreshing session']
    );
    expect(hardwareScalingStatus).toMatchSnapshot('Refreshing session');
    await mockGetUtilizationData();
    await immediatePromise();
    expect(mockOnComplete).toBeCalled();
    expect(hardwareScalingStatus.state('status')).toEqual(Status.Complete);
    expect(hardwareScalingStatus).toMatchSnapshot('Complete');
  });

  it('Renders with auth error', async () => {
    const rejectedAuth = Promise.reject('No token');
    mockAuthTokenRetrieval.mockReturnValue(rejectedAuth);
    const detailsResponse = JSON.parse(DETAILS_RESPONSE);
    const hardwareConfig = detailsToHardwareConfiguration(detailsResponse);
    const hardwareScalingStatus = shallow(
      <HardwareScalingStatus
        detailsServer={mockServerWrapper}
        notebookService={mockNotebookService}
        onCompletion={mockOnComplete}
        onDialogClose={mockOnDialogClose}
        hardwareConfiguration={hardwareConfig}
        authTokenRetrieval={mockAuthTokenRetrieval}
        machineTypes={mockMachineTypes}
      />
    );
    await rejectedAuth.catch(() => {});
    expect(hardwareScalingStatus.state('status')).toEqual(Status.Error);
    expect(hardwareScalingStatus).toMatchSnapshot('Auth Error');
  });

  it('Renders with stop error', async () => {
    mockAuthTokenRetrieval.mockResolvedValue(mockToken);
    const rejectedStopOperation = Promise.reject('Stop operation failed');
    mockStopInstance.mockReturnValue(rejectedStopOperation);
    const detailsResponse = JSON.parse(DETAILS_RESPONSE);
    const hardwareConfig = detailsToHardwareConfiguration(detailsResponse);
    const hardwareScalingStatus = shallow(
      <HardwareScalingStatus
        detailsServer={mockServerWrapper}
        notebookService={mockNotebookService}
        onCompletion={mockOnComplete}
        onDialogClose={mockOnDialogClose}
        hardwareConfiguration={hardwareConfig}
        authTokenRetrieval={mockAuthTokenRetrieval}
        machineTypes={mockMachineTypes}
      />
    );
    await mockAuthTokenRetrieval();
    await rejectedStopOperation.catch(() => {});
    expect(hardwareScalingStatus.state('status')).toEqual(Status.Error);
    expect(hardwareScalingStatus).toMatchSnapshot('Stop instance error');
  });

  it('Renders with operation error', async () => {
    const mockGetUtilizationData = jest.fn();
    const mockStopInstance = jest.fn();
    const rejectedSetMachine = Promise.reject('Reshape operation failed');
    const rejectedSetAccelerator = Promise.reject('Reshape operation failed');
    mockSetMachineType.mockReturnValue(rejectedSetMachine);
    mockSetAccelerator.mockReturnValue(rejectedSetAccelerator);
    mockStartInstance.mockResolvedValue(mockInstance);
    const detailsResponse = JSON.parse(DETAILS_RESPONSE);
    const resolveValue = Promise.resolve({ ...detailsResponse });
    mockGetUtilizationData.mockReturnValue(resolveValue);
    const hardwareConfig = detailsToHardwareConfiguration(detailsResponse);
    const hardwareScalingStatus = shallow(
      <HardwareScalingStatus
        detailsServer={mockServerWrapper}
        notebookService={mockNotebookService}
        onCompletion={mockOnComplete}
        onDialogClose={mockOnDialogClose}
        hardwareConfiguration={hardwareConfig}
        authTokenRetrieval={mockAuthTokenRetrieval}
        machineTypes={mockMachineTypes}
      />
    );
    await mockAuthTokenRetrieval();
    await mockStopInstance();
    await rejectedSetMachine.catch(() => {});
    await rejectedSetAccelerator.catch(() => {});
    await mockStartInstance();
    await immediatePromise();
    expect(hardwareScalingStatus.state('status')).toEqual(Status.Error);
    expect(hardwareScalingStatus).toMatchSnapshot('Operation error');
  });

  it('Renders with restart error', async () => {
    const mockGetUtilizationData = jest.fn();
    const mockStopInstance = jest.fn();
    const rejectedSetMachine = Promise.reject('Reshape operation failed');
    const rejectedSetAccelerator = Promise.reject('Reshape operation failed');
    const rejectedStartOperation = Promise.reject('Restart operation failed');
    mockSetMachineType.mockReturnValue(rejectedSetMachine);
    mockSetAccelerator.mockReturnValue(rejectedSetAccelerator);
    mockStartInstance.mockRejectedValue(rejectedStartOperation);
    const detailsResponse = JSON.parse(DETAILS_RESPONSE);
    const resolveValue = Promise.resolve({ ...detailsResponse });
    mockGetUtilizationData.mockReturnValue(resolveValue);
    const hardwareConfig = detailsToHardwareConfiguration(detailsResponse);
    const hardwareScalingStatus = shallow(
      <HardwareScalingStatus
        detailsServer={mockServerWrapper}
        notebookService={mockNotebookService}
        onCompletion={mockOnComplete}
        onDialogClose={mockOnDialogClose}
        hardwareConfiguration={hardwareConfig}
        authTokenRetrieval={mockAuthTokenRetrieval}
        machineTypes={mockMachineTypes}
      />
    );
    await mockAuthTokenRetrieval();
    await mockStopInstance();
    await rejectedSetMachine.catch(() => {});
    await rejectedSetAccelerator.catch(() => {});
    await rejectedStartOperation.catch(() => {});
    expect(hardwareScalingStatus.state('status')).toEqual(Status.Error);
    expect(hardwareScalingStatus).toMatchSnapshot('Restart error');
  });
});
