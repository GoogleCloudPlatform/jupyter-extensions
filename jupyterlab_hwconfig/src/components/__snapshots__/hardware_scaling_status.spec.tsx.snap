// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HardwareScalingStatus Renders with auth error: Auth Error 1`] = `
<ErrorPage
  error={null}
  instanceDetails={null}
  machineTypes={
    Array [
      Object {
        "base": Object {
          "text": "Efficient Instance",
          "value": "e2-",
        },
        "configurations": Array [
          Object {
            "text": "Efficient Instance, 8 vCPUs, 64 GB RAM",
            "value": "e2-highmem-8",
          },
        ],
      },
    ]
  }
  onDialogClose={[MockFunction]}
/>
`;

exports[`HardwareScalingStatus Renders with operation error: Operation error 1`] = `
<ErrorPage
  error={
    Object {
      "errorType": "Reshape",
      "errorValue": "Reshape operation failed",
    }
  }
  instanceDetails={
    Object {
      "acceleratorConfig": Object {
        "coreCount": "1",
        "type": "NVIDIA_TESLA_K80",
      },
      "machineType": "https://www.googleapis.com/compute/v1/projects/jupyterlab-interns-sandbox/zones/us-west1-b/machineTypes/e2-highmem-8",
      "name": "Test",
    }
  }
  machineTypes={
    Array [
      Object {
        "base": Object {
          "text": "Efficient Instance",
          "value": "e2-",
        },
        "configurations": Array [
          Object {
            "text": "Efficient Instance, 8 vCPUs, 64 GB RAM",
            "value": "e2-highmem-8",
          },
        ],
      },
    ]
  }
  onDialogClose={[MockFunction]}
/>
`;

exports[`HardwareScalingStatus Renders with restart error: Restart error 1`] = `
<ErrorPage
  error={
    Object {
      "errorType": "Start",
      "errorValue": Promise {},
    }
  }
  machineTypes={
    Array [
      Object {
        "base": Object {
          "text": "Efficient Instance",
          "value": "e2-",
        },
        "configurations": Array [
          Object {
            "text": "Efficient Instance, 8 vCPUs, 64 GB RAM",
            "value": "e2-highmem-8",
          },
        ],
      },
    ]
  }
  onDialogClose={[MockFunction]}
/>
`;

exports[`HardwareScalingStatus Renders with stop error: Stop instance error 1`] = `
<ErrorPage
  error={
    Object {
      "errorType": "Stop",
      "errorValue": "Stop operation failed",
    }
  }
  instanceDetails={null}
  machineTypes={
    Array [
      Object {
        "base": Object {
          "text": "Efficient Instance",
          "value": "e2-",
        },
        "configurations": Array [
          Object {
            "text": "Efficient Instance, 8 vCPUs, 64 GB RAM",
            "value": "e2-highmem-8",
          },
        ],
      },
    ]
  }
  onDialogClose={[MockFunction]}
/>
`;

exports[`HardwareScalingStatus Runs through reshaping flow: Authorizing 1`] = `
<div
  className="containerPadding_f7mmknj container_f11ewvlg"
>
  <span
    className="image_fc8jh0z jp-CaipIcon"
  />
  <p
    className="heading_f1xburhd"
  >
    Authorizing (1/6)
  </p>
  <div
    className="container_f11ewvlg"
  >
    <WithStyles(WithStyles(ForwardRef(LinearProgress)))
      value={16.666666666666664}
      variant="determinate"
    />
    <p
      className="paragraph_f1ditx77"
    >
      Please complete the OAuth 2.0 authorization steps in the popup
    </p>
    <p
      className="bottomText_f1vi2ddg"
    >
      Don't close the browser tab before the update process is finished
    </p>
  </div>
</div>
`;

exports[`HardwareScalingStatus Runs through reshaping flow: Complete 1`] = `
<div
  className="containerPadding_f7mmknj container_f11ewvlg"
>
  <span
    className="image_fc8jh0z jp-CaipIcon"
  />
  <p
    className="heading_f1xburhd"
  >
    Complete
  </p>
  <div
    className="containerBottomPadding_f1uhfe4m"
  >
    <p
      className="paragraph_f1ditx77"
    >
      Operation complete. Enjoy your newly configured instance!
    </p>
    <div>
      <span
        className="subheading_fgjqs4u"
      >
        Your new configuration:
      </span>
      <div
        className="paragraph_f1ditx77"
      >
        Machine type: 
        Efficient Instance, 8 vCPUs, 64 GB RAM
      </div>
      <div
        className="paragraph_f1ditx77"
      >
        GPUs: 1 NVIDIA Tesla K80
      </div>
    </div>
    <ActionBar
      onPrimaryClick={[MockFunction]}
      primaryLabel="Close"
    />
  </div>
</div>
`;

exports[`HardwareScalingStatus Runs through reshaping flow: Refreshing session 1`] = `
<div
  className="containerPadding_f7mmknj container_f11ewvlg"
>
  <span
    className="image_fc8jh0z jp-CaipIcon"
  />
  <p
    className="heading_f1xburhd"
  >
    Refreshing session (6/6)
  </p>
  <div
    className="container_f11ewvlg"
  >
    <WithStyles(WithStyles(ForwardRef(LinearProgress)))
      value={100}
      variant="determinate"
    />
    <p
      className="paragraph_f1ditx77"
    >
      Refreshing your JupyterLab session to reflect your new configuration.
    </p>
    <p
      className="bottomText_f1vi2ddg"
    >
      Don't close the browser tab before the update process is finished
    </p>
  </div>
</div>
`;

exports[`HardwareScalingStatus Runs through reshaping flow: Restarting notebook instance 1`] = `
<div
  className="containerPadding_f7mmknj container_f11ewvlg"
>
  <span
    className="image_fc8jh0z jp-CaipIcon"
  />
  <p
    className="heading_f1xburhd"
  >
    Restarting notebook instance (5/6)
  </p>
  <div
    className="container_f11ewvlg"
  >
    <WithStyles(WithStyles(ForwardRef(LinearProgress)))
      value={83.33333333333334}
      variant="determinate"
    />
    <p
      className="paragraph_f1ditx77"
    >
      Restarting your newly configured notebook instance.
    </p>
    <p
      className="bottomText_f1vi2ddg"
    >
      Don't close the browser tab before the update process is finished
    </p>
  </div>
</div>
`;

exports[`HardwareScalingStatus Runs through reshaping flow: Stopping notebook instance 1`] = `
<div
  className="containerPadding_f7mmknj container_f11ewvlg"
>
  <span
    className="image_fc8jh0z jp-CaipIcon"
  />
  <p
    className="heading_f1xburhd"
  >
    Stopping notebook instance (2/6)
  </p>
  <div
    className="container_f11ewvlg"
  >
    <WithStyles(WithStyles(ForwardRef(LinearProgress)))
      value={33.33333333333333}
      variant="determinate"
    />
    <p
      className="paragraph_f1ditx77"
    >
      Shutting down your notebook instance for reshaping.
    </p>
    <p
      className="bottomText_f1vi2ddg"
    >
      Don't close the browser tab before the update process is finished
    </p>
  </div>
</div>
`;

exports[`HardwareScalingStatus Runs through reshaping flow: Updating GPU configuration 1`] = `
<div
  className="containerPadding_f7mmknj container_f11ewvlg"
>
  <span
    className="image_fc8jh0z jp-CaipIcon"
  />
  <p
    className="heading_f1xburhd"
  >
    Updating GPU configuration (4/6)
  </p>
  <div
    className="container_f11ewvlg"
  >
    <WithStyles(WithStyles(ForwardRef(LinearProgress)))
      value={66.66666666666666}
      variant="determinate"
    />
    <p
      className="paragraph_f1ditx77"
    >
      Reshaping GPU configuration to match your selection.
    </p>
    <p
      className="bottomText_f1vi2ddg"
    >
      Don't close the browser tab before the update process is finished
    </p>
  </div>
</div>
`;

exports[`HardwareScalingStatus Runs through reshaping flow: Updating machine configuration 1`] = `
<div
  className="containerPadding_f7mmknj container_f11ewvlg"
>
  <span
    className="image_fc8jh0z jp-CaipIcon"
  />
  <p
    className="heading_f1xburhd"
  >
    Updating machine configuration (3/6)
  </p>
  <div
    className="container_f11ewvlg"
  >
    <WithStyles(WithStyles(ForwardRef(LinearProgress)))
      value={50}
      variant="determinate"
    />
    <p
      className="paragraph_f1ditx77"
    >
      Reshaping machine configuration to match your selection.
    </p>
    <p
      className="bottomText_f1vi2ddg"
    >
      Don't close the browser tab before the update process is finished
    </p>
  </div>
</div>
`;
